### ISSUE TITLE
will it work for  multivariate  time series classification for example mixture of categorical and continues data? 

### ISSUE THREAD

**Issue by Sandy4321** [ymtoo/dtw-dist#1]
will it work for  multivariate  time series classification for example mixture of categorical and continues data? 
	for example at time t1 we have observation: red, 2.4 , 5, 12.456 and time t2: green, 3.5, 2, 45.78; time t3: black, 5.6, 7, 23.56; t4: red, 2.1, 5, 12.6 ?

**Comment by ymtoo** [comment]
To measure the distance between two multivariate time series data which consist of categorical and numerical data, you have to convert the categorical data into numerical forms. The following example could be one of the potential solutions.

```python
from matplotlib import colors
import numpy as np

from ts_dist import dtw_dist as dtw_dist_py

def get_rgb(X_color):
    """Convert colors to RGB values."""
    if not isinstance(X_color, (list, tuple)):
        X_color = list(X_color)
    n_time = len(X_color)
    rgb = np.zeros((3, n_time))
    for i in range(n_time):
        rgb[:, i] = colors.to_rgb(X_color[i])
    return rgb

X = [["red", "green", "black", "red"], 
     [2.4, 3.5, 5.6, 2.1], 
     [5, 2, 7, 5], 
     [12.456, 45.78, 23.56, 12.6]]
Y = [["yellow", "white", "white", "red"], 
     [2.4, 3.5, 5.6, 2.1], 
     [5, 2, 7, 5], 
     [12.456, 45.78, 23.56, 12.6]]

X_color = get_rgb(X[0])
X_mod = np.concatenate((X_color, np.array(X[1:])))
Y_color = get_rgb(Y[0])
Y_mod = np.concatenate((Y_color, np.array(Y[1:])))

dtw_py = dtw_dist_py(X_mod, Y_mod)

```

**Comment by Sandy4321** [comment]
I see thanks
But what is about scale
Some dimension is around 1
Other are around 20
What distance metric you use
Is it something like Euclid metric?

**Comment by ymtoo** [comment]
You can normalize/standardize the data accordingly.

```Python
X_color[:3, :] = X_mod[:3, :]/1.0 #maximum value of colors is 1.0
X_color[3:, :] = X_mod[3:, :]/X_mod.max(axis=1, keepdims=True) #do not know the maximum value
Y_color[:3, :] = Y_mod[:3, :]/1.0 #maximum value of colors is 1.0
Y_color[3:, :] = Y_mod[3:, :]/Y_mod.max(axis=1, keepdims=True) #do not know the maximum value
```

The supported distance measures are

* Dynamic Time Warping (DTW)
* Longest Common Subsequence (LCSS)
* Edit Distance on Real sequence (EDR). 

Unlike the Euclidean metric, these distance measures allow non-linear mapping between two time series data which may vary in timing.

**Comment by Sandy4321** [comment]
great thanks, will it work for multidimensional time series's ?

**Comment by ymtoo** [comment]
Yes. In fact, the example above is dealing with 4-D data. The number of rows of input arguments (X, Y) is the number of dimensions.

**Comment by Sandy4321** [comment]
This link is about this
https://link.springer.com/article/10.1007/s10618-016-0455-0

**Comment by Sandy4321** [comment]
Generalizing DTW to the multi-dimensional case requires an adaptive approach

**Comment by ymtoo** [comment]
The current implementation (dependent and independent DTW) is non-adaptive. A quick scan through the paper, it seems like the adaptive part of the algorithm requires a classifier and a training dataset. The aim of this module is to solely provide functions for distance measures between two time-series data. You can use this module to compute the dependent and independent DTW in the adaptive classification algorithm in the paper. Interesting paper though!

**Comment by Sandy4321** [comment]
great keep me updated for new code
